images/egress/dns-proxy/egress-dns-proxy.sh

```diff
@@ -0,0 +1,183 @@
+    maxconn     4000
+    user        haproxy
+    group       haproxy
+
+defaults
+    log                     global
+    mode                    tcp
+    option                  dontlognull
+    option                  tcplog
+    option                  redispatch
+    retries                 3
+    timeout http-request    100s
+    timeout queue           1m
+    timeout connect         10s
+    timeout client          1m
+    timeout server          1m
+    timeout http-keep-alive 100s
+    timeout check           10s
+"
+}
+
+function generate_dns_resolvers() {
+    echo "resolvers dns-resolver"
+    # Fetch nameservers from /etc/resolv.conf
+    declare -a nameservers=$(cat /etc/resolv.conf |grep nameserver|awk -F" " '{print $2}')
+    n=0
+    for ns in ${nameservers[@]}; do
+        n=$(($n + 1))
+        echo "    nameserver ns$n ${ns}:${NS_PORT}"
+    done
+
+    # Add google DNS servers as fallback
+    echo "    nameserver nsfallback1 8.8.8.8:${NS_PORT}"
+    echo "    nameserver nsfallback2 8.8.4.4:${NS_PORT}"
+
+    # Set default optional params
+    echo "    resolve_retries      3"
+    echo "    timeout retry        1s"
+    echo "    hold valid           10s"
+    echo ""
+}
+
+function generate_haproxy_frontends_backends() {
+    local n=0
+    declare -A used_ports=()
+
+    while read dest; do
+        local port target targetport resolvers 
+
+        if [[ "${dest}" =~ ^${BLANK_LINE_OR_COMMENT_REGEX}$ ]]; then
+            continue
+        fi
+        n=$(($n + 1))
+        resolvers=""
+
+        if [[ "${dest}" =~ ^${PORT_REGEX}\ +${IPADDR_REGEX}$ ]]; then
+            read port target <<< "${dest}"
+            targetport="${port}"
+        elif [[ "${dest}" =~ ^${PORT_REGEX}\ +${IPADDR_REGEX}\ +${PORT_REGEX}$ ]]; then
+            read port target targetport <<< "${dest}"
+        elif [[ "${dest}" =~ ^${PORT_REGEX}\ +${DOMAINNAME_REGEX}$ ]]; then
+            read port target <<< "${dest}"
+            targetport="${port}"
+            resolvers="resolvers dns-resolver"
+        elif [[ "${dest}" =~ ^${PORT_REGEX}\ +${DOMAINNAME_REGEX}\ +${PORT_REGEX}$ ]]; then
+            read port target targetport <<< "${dest}"
+            resolvers="resolvers dns-resolver"
+        else
+            die "Bad destination '${dest}'"
+        fi
+
+        validate_port ${port}
+        validate_port ${targetport}
+
+        if [[ "${used_ports[${port}]:-}" == "" ]]; then
+            used_ports[${port}]=1
+        else
+            die "Proxy port $port already used, must be unique for each destination"
+        fi
+
+        echo "
+frontend fe$n
+    bind ${EGRESS_SOURCE}:${port}
+    default_backend be$n
+
+backend be$n
+    server dest$n ${target}:${targetport} check $resolvers
+"
+    done <<< "${EGRESS_DESTINATION}"
+}
+
+function setup_haproxy_config() {
+    generate_haproxy_defaults
+    generate_dns_resolvers
+    generate_haproxy_frontends_backends
+}
+
+function setup_haproxy_syslog() {
```

Comment:
Ugh. There's no way to get it to just log to stdout/stderr?

We might want to make it not bother doing this unless the pod spec sets some sort of "debug" variable...